### ISSUE TITLE
Comments should be AST nodes

### ISSUE THREAD

**Issue by 1ucian0** [Qiskit/qiskit-terra#901]
Consider the following program:
```
from qiskit.qasm import Qasm
qasm_string = """//this is a comment
OPENQASM 2.0;
qreg q[2]; // This is another comment
"""

node_circuit = Qasm(data=qasm_string).parse()
print(node_circuit.qasm())
print([ type(i) for i in node_circuit.children])
```
The output is:
```
OPENQASM 2.0;
qreg q[2];

[<class 'qiskit.qasm._node._format.Format'>, <class 'qiskit.qasm._node._qreg.Qreg'>]
```

### What is the expected enhancement?

The following output:
```
//this is a comment
OPENQASM 2.0; 
qreg q[2];
// This is another comment

[<class 'qiskit.qasm._node._comment.Comment'>, <class 'qiskit.qasm._node._format.Format'>, <class 'qiskit.qasm._node._qreg.Qreg'>, <class 'qiskit.qasm._node._comment.Comment'>]
```

The current parser ignores comments. There is room for discussion if a parser should create an AST with this information (a Full Syntax Tree).  I think in our context is a handy thing to have. For example, to add annotations or metainformation, which might be handy in such experimental field.

**Comment by delapuente** [comment]
Usually, comments are ignored in the lexer stage, but I'm a big fan of comment nodes. A common strategy is to provide a flag for retaining these nodes or not.

**Comment by delapuente** [comment]
I'm not sure we the expected behaviour is what you point. I would have expected the new code to keep the comments in the same locations, aside from some normalization so this:

```
OPENQASM 2.0;                          // a very separated comment
```

would translate to this:

```
OPENQASM 2.0; // a very separated comment
```

But it looks counter-intuitive to find the comment in a new line after the code it is commenting. Even thinking that the behaviour preserves the ordering of the nodes, it challenges the semantic purpose of the comment. Suppose:

```
OPENQASM 2.0;
if (c0 == 1) cnot(q[0], q[1]);  // it depends on the first classical register
if (c1 == 1) cnot(q[0], q[1]);
```

would be translated into:

```
OPENQASM 2.0;
if (c0 == 1) cnot(q[0], q[1]);
// it depends on the first classical register
if (c1 == 1) cnot(q[0], q[1]);
```

which makes the comment to refer to the last statement, which is not true.

I would rather expect this comment to be placed before the statement but the optimal solution is to keep it at the statement level.

The OPEN QASM 2.0 spec is inconsistent regarding comments: the [formal grammar](https://github.com/Qiskit/openqasm/blob/master/spec/qasm2.rst#open-qasm-grammar) does not include them (suggesting they are unimportant from a semantic point of view) but a summarizing statements table at the end of the [language description section]((https://github.com/Qiskit/openqasm/blob/master/spec/qasm2.rst#language)) include comments among the statements.

From a grammatical point of view, one of the solutions is to retain the concrete the [concrete syntax tree or parsing tree](https://en.wikipedia.org/wiki/Parse_tree) and use it to recompose the original form of the source code.

Another, perhaps more practical solution, would be to allow the AST nodes to carry some metainformation that can include an inlinded comment node.

**Comment by 1ucian0** [comment]
The "unparse" method is `qasm()` and it has some room for improvement in terms of reconstructing line and spacing or using the positioning metadata.
Examples:
<table align="left">
  <tr>
    <th> input to parse </th>
    <th> output of unparse</th>
    </tr>
  <tr>
    <td>
<pre>
OPENQASM 2.0;
qreg q[2]; qreg r[1];
</pre>
   </td>
    <td>
<pre>
OPENQASM 2.0;
qreg q[2];
qreg r[1];
</pre>
    </td>
    </tr>

  <tr>
    <td>
<pre>
OPENQASM 2.0;
qreg q[2];




&nbsp;
qreg r[1];</pre>
   </td>
    <td>
<pre>
OPENQASM 2.0;
qreg q[2];
qreg r[1];
</pre>
    </td>
    </tr>


  <tr>
    <td>
<pre>
OPENQASM 2.0;
qreg q[2];                           qreg r[1];</pre>
   </td>
    <td>
<pre>
OPENQASM 2.0;
qreg q[2];
qreg r[1];
</pre>
    </td>
    </tr>

<tr>
    <td>
<pre>
OPENQASM 2.0;
gate name q {}   </td>
    <td>
<pre>OPENQASM 2.0;
gate name q
{
}
</pre>
    </td>
    </tr>
</table>

Basically, the (un)parser is ignoring spacing and line positioning and my implementation of comment nodes (PR #906) follows that previous philosophy.

**Comment by 1ucian0** [comment]
In your example:
```
if (c0 == 1) cnot(q[0], q[1]);
// it depends on the first classical register
if (c1 == 1) cnot(q[0], q[1]);
```
the reason for this is not connected with the comment. The new line between the comment and the conditional statement is added by the state concatenation code (see https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/qasm/_node/_program.py#L28).

**Comment by delapuente** [comment]
As mentioned before, I don't mind some normalization happening. The output of the unparsing is more verbose, agree, but it is not incorrect. However, making an inline comment to be pushed after the code commented is against a very well-established tradition of putting comments before or at the same level of the commented code.

I don't mind to keep this open to track the fract that OPEANQASM 2.0 has a documented case where comments are relevant for the AST as docstrings that we are not implementing. But the current description of the issue is incorrect, even talking about ASTs only:

![captura de pantalla 2018-09-21 a las 20 10 00](https://user-images.githubusercontent.com/757942/45898490-6502c280-bdda-11e8-9c9f-9d6fbf8de871.png)

**Comment by delapuente** [comment]
It is incorrect because something like:
```
OPENQASM 2.0;
qreg q[2]; // qreg comment
gate custom q { ... }
```
would be parsed as: `[Qreg, Comment, Gate]`.
And:
```
OPENQASM 2.0;
qreg q[2];
// Metadata for custom gate
gate custom q { ... }
```
would be equally parsed: `[Qreg, Comment, Gate]` without a way of distinguishing among them.

**Comment by 1ucian0** [comment]
The unparser if patched in ff1f735, by keeping line information.

**Comment by delapuente** [comment]
@1ucian0 how do you suggest the AST should look when parsing an inline comment?

**Comment by 1ucian0** [comment]
after ff1f735, the line information is saved and the unparse puts the inline comment in the correct line. Currently the parse parse this:
```
OPENQASM 2.0;
qreg q[2];
// Metadata for custom gate
gate custom q { ... }
```
into something like this:
```
prog:
  |- header - number
  |- comment
  |- gate ...
```

Changing it into this:
```
prog:
  |- header - number
  |- gate ...
       |- comment
```
is not trivial.

**Comment by chetmurthy** [comment]
I think this is going to be very difficult to do in a sensible way.  Consider the code

```CX // argle
q[0],q[1];
```

This is valid, and parses the same as 

```CX q[0],q[1];
```

To make sufficient changes to the AST to support comments basically *anywhere* that are part of the AST, is going to make it insanely ugly.  And it'll make the grammar a real mess, to boot.

But there is hope.  There is a more-or-less standard way of getting what you wish, at the cost of a smarter lexer.  Basically, instead of the lexer returning tokens that are mere strings (or other constants), it returns a triple, of (<token>, <start-position in input>, <end-position in input>).  Both positions are returned, so that (fixed-width) operators and keywords can be handled the same as identifiers and other variable-width tokens.  The AST node has a slot where the tokens returned in parsing that AST node *directly* (not its children) are stored.  And at pretty-printing time, instead of (possibly empty) whitespace between text, the pretty-printer uses the saved tokens to look at the original input stream and pull out the whitespace (if any) which was used originally, and prints that.

I don't know if PLY supports this mode of use; I'll look into it.  I know that Ocaml's "alternate" parser/lexer system, Camlp4/Camlp5 supports it -- it was a design point -- and I suspect that the standard parser/lexer can also be used in that way.

But basically, saving the comments in the AST is .... not a good idea.

**Comment by 1ucian0** [comment]
What do you mean by `q[0],q[1];` being valid?

It's my impression that the OpenQASM authors thought this way to add "metadata" to the gate definition:
```
// comment
gate name(params) qargs
{
body
```
I agree that might not be the best way to reach their goal. This kind of comment might be treat differently than `h q[0]; \\ comment` by the parser. I don't have any use case for their later use. But I would like to support the former.

**Comment by chetmurthy** [comment]
Arrrgh!  I didn't even *notice* that github took out the comment.  What I meant was that comments are valid anywhere, e.g. between the "," and the "q" in "cx q[0], q[1] ;"

Comments are valid between any two lexemes, and if you want a comment to be a token, then the modifications to the grammar to support that will be .... *painful*.  Whereas the method I suggest is both standard, and straightforward to implement.

**Comment by 1ucian0** [comment]
My understanding of the specification is that comments are between `//` and line-end.

Probably my original post was wrongly trying to be too generic. Sorry. I'm only interested in the use case from the specification, about adding metadata to gates.

**Comment by chetmurthy** [comment]
Comments *are* between `//` and line-end.

**Comment by 1ucian0** [comment]
oh! I see what you mean:
```
OPENQASM 2.0;
qreg q[2]; 
CX q[0], // comment 
q[1]; 
```
You are right.

**Comment by jwoehr** [comment]
An end-of-line comment is one thing; inserting a comment after the comma, splitting the line and naming a target register on the next line looks quite odd.

Those cases look pathological. I think no classical assembler would accept that syntax?

**Comment by jwoehr** [comment]
That said, comments are indeed nodes in nuqasm2, but I have not dealt with the case of an end-of-line comment, which easily could be done.

**Comment by chetmurthy** [comment]
I wasn't suggesting that such comments are desired; rather, that if you want to make comments be part of the grammar, you're going to have to change where they're allowed, b/c currently such pathological cases are allowed.  That's all I mean.

Also, you're probably right that "old-school line-oriented language-processors" probably don't accept such comments.  And I find that GNU AS doesn't accept them, either. 

The problem is easily solvable -- just prohibit comments except at end-of-statement (and at beginning-of-file).

**Comment by jwoehr** [comment]
Currently my nuqasm2 POC accepts comment lines anywhere and adds them to the AST. However, I didn't consider the end-of-line comment, which probably should be permitted.
